### PR TITLE
fix: fix number mask bug with wrong position of thousand separator

### PR DESCRIPTION
**Pull request by Mahakon**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [ ] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
-   [ ] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

-   [X] Bugfix
-   [ ] Feature
-   [ ] Refactoring
-   [ ] Code style update
-   [ ] Build or CI related changes
-   [ ] Documentation content changes

## What is the current behavior?

https://github.com/TinkoffCreditSystems/taiga-ui/issues/984

## What is the new behavior?

fix bug

## Does this PR introduce a breaking change?

-   [] Yes
-   [X] No

### CODE REVIEW

File: projects/core/utils/mask/create-number-mask.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export function tuiCreateNumberMask({
             : integer;
         const integerCappedClean = integerCapped
             .replace(TUI_NON_DIGITS_REGEXP, '')
-            .replace(/^0+(?!\.|$)/, '0');
+            .replace(/^0+(?!\.|$)/, '');
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:78]: add tests please for this case))

File: projects/core/utils/mask/create-number-mask.ts [hunk 2]
```diff
@@ -115,7 +115,11 @@ export function tuiCreateNumberMask({
 function preventLeadingZeroes(mask: Array<string | RegExp>): Array<string | RegExp> {
     const firstDigitIndex = mask.indexOf(TUI_DIGIT_REGEXP);
 
-    if (firstDigitIndex !== -1 && mask[firstDigitIndex + 1] === TUI_DIGIT_REGEXP) {
+    if (
+        firstDigitIndex !== -1 &&
+        mask[firstDigitIndex] === TUI_DIGIT_REGEXP &&
+        mask.length !== 1
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:121]: maybe move it to const variable and use

```ts
if (condition) {
  mask[firstDigitIndex] = NON_ZERO_DIGIT;
}

File: projects/core/utils/mask/create-number-mask.ts [hunk 3]
```diff
@@ -115,7 +115,11 @@ export function tuiCreateNumberMask({
 function preventLeadingZeroes(mask: Array<string | RegExp>): Array<string | RegExp> {
     const firstDigitIndex = mask.indexOf(TUI_DIGIT_REGEXP);
 
-    if (firstDigitIndex !== -1 && mask[firstDigitIndex + 1] === TUI_DIGIT_REGEXP) {
+    if (
+        firstDigitIndex !== -1 &&
+        mask[firstDigitIndex] === TUI_DIGIT_REGEXP &&
```
Comment (waterplea) [inline projects/core/utils/mask/create-number-mask.ts:120]: This is a useless check, we already know it's true because this is how we got `firstDigitIndex` in the first place. From what I understand, we check that there's a digit in the mask and there's another digit after it. You changed it to just see if there is a digit in the mask and the mask is not one symbol. Could you elaborate on this change? Now that it doesn't look for two digits in a row, could it have issues with decimal part separator?

File: projects/core/utils/mask/create-number-mask.ts (hunk 1 above)
Comment (waterplea) [reply projects/core/utils/mask/create-number-mask.ts:78]: Struggling to understand what this line does :D

File: projects/core/utils/mask/create-number-mask.ts [hunk 4]
```diff
@@ -108,14 +108,29 @@ export function tuiCreateNumberMask({
...
+    if (firstDigitIndex === -1) {
+        return mask;
+    }
+
+    const isOnlyNonZeroDigit =
+        mask.length === 1 && NON_ZERO_DIGIT.test(integerCappedClean);
+
+    if (
+        isOnlyNonZeroDigit ||
+        [TUI_DIGIT_REGEXP, thousandSymbol].includes(mask[firstDigitIndex + 1])
```
Comment (waterplea) [inline projects/core/utils/mask/create-number-mask.ts:132]: We do not use `includes` on arrays, in this major version we stick to es6 code. When we bump supported browser versions in 3.0, I think we will start using it, but for now we stick to `indexOf`.

File: projects/core/utils/mask/create-number-mask.ts (hunk 1 above)
Comment (waterplea) [reply projects/core/utils/mask/create-number-mask.ts:78]: @Mahakon can you clarify what this line was doing and what is the new behavior? It's been a while since I looked into this code.

File: projects/core/utils/mask/create-number-mask.ts [hunk 5]
```diff
@@ -108,14 +108,29 @@ export function tuiCreateNumberMask({
             mask.unshift('-');
         }
 
-        return preventLeadingZeroes(mask);
+        return preventLeadingZeroes(mask, thousandSymbol, integerCappedClean);
+        //return mask;
```
Comment (waterplea) [inline projects/core/utils/mask/create-number-mask.ts:112]: Please remove commented code.

File: projects/core/utils/mask/create-number-mask.ts (hunk 1 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: so here is leaved one zero as a flag to add NON_ZERO_DIGIT prevent zeroes

File: projects/core/utils/mask/create-number-mask.ts (hunk 4 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: refactored this part

File: projects/core/utils/mask/create-number-mask.ts (hunk 5 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: done

File: projects/core/utils/mask/create-number-mask.ts (hunk 1 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: because of this extra symbol it do not write separate number with ThousandsSeparator

File: projects/core/utils/mask/create-number-mask.ts [hunk 6]
```diff
@@ -73,10 +74,16 @@ export function tuiCreateNumberMask({
...
             : integer;
-        const integerCappedClean = integerCapped
-            .replace(TUI_NON_DIGITS_REGEXP, '')
-            .replace(/^0+(?!\.|$)/, '0');
-        const withSeparator = addThousandsSeparator(integerCappedClean, thousandSymbol);
+        const integerCappedClean = integerCapped.replace(TUI_NON_DIGITS_REGEXP, '');
+        const leadingZerosMarchArray =
+            integerCappedClean.match(TUI_LEADING_ZEROES_REGEXP) || [];
+        const leadingZerosAmount =
+            leadingZerosMarchArray.length !== 0 ? leadingZerosMarchArray[0].length : 0;
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:81]:
```ts
const leadingZerosMarchArray =
            integerCappedClean.match(TUI_LEADING_ZEROES_REGEXP) || [];
const leadingZerosAmount =
            leadingZerosMarchArray.length !== 0 ? leadingZerosMarchArray[0].length : 0;
```

```ts
const [ leadingZerosAmount ] = integerCappedClean.match(TUI_LEADING_ZEROES_REGEXP) || [ 0 ];
```

File: projects/core/utils/mask/create-number-mask.ts [hunk 7]
```diff
@@ -73,10 +74,16 @@ export function tuiCreateNumberMask({
...
+        const integerCappedZerosClean = integerCappedClean.replace(/^0+(?!\.|$)/, '');
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:82]: do you need a `trim()` method of the end `.replace(/^0+(?!\.|$)/, '')`?

File: projects/core/utils/mask/create-number-mask.ts [hunk 8]
```diff
@@ -108,17 +117,37 @@ export function tuiCreateNumberMask({
...
+function preventLeadingZeroes(
+    mask: Array<string | RegExp>,
+    isOnlyZeroDigit: boolean = false,
+    leadingZerosAmount: number = 0,
+): Array<string | RegExp> {
+    if (isOnlyZeroDigit || leadingZerosAmount < 0) {
+        return mask;
+    }
+
     const firstDigitIndex = mask.indexOf(TUI_DIGIT_REGEXP);
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:134]: what do you think about?

```ts
const secondMaskDigit = mask[firstDigitIndex + 1];
const isCaretTrap = secondMaskDigit === MASK_CARET_TRAP;

if (isCaretTrap && leadingZerosAmount === 1) {
        return mask;
} else if (isCaretTrap) {
        mask.unshift(NON_ZERO_DIGIT);
        return mask;
}
```

File: projects/kit/components/input-number/test/input-number.component.spec.ts [hunk 9]
```diff
@@ -270,6 +270,22 @@ describe('InputNumber', () => {
             expect(component.computedValue).toBe(`-12${CHAR_NO_BREAK_SPACE}345,67`);
         });
 
+        it(`The given value from the form is correctly converted to a string with 0's at start`, () => {
```
Comment (splincode) [inline projects/kit/components/input-number/test/input-number.component.spec.ts:273]: Are there any other edge cases worth considering in the tests?

File: projects/core/utils/mask/create-number-mask.ts (hunk 6 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: refactored

File: projects/core/utils/mask/create-number-mask.ts (hunk 7 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: added

File: projects/core/utils/mask/create-number-mask.ts (hunk 8 above)
Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts:136]: done

File: projects/kit/components/input-number/test/input-number.component.spec.ts (hunk 9 above)
Comment (Mahakon) [reply projects/kit/components/input-number/test/input-number.component.spec.ts:273]: add one more

File: projects/core/utils/mask/create-number-mask.ts [hunk 10]
```diff
@@ -75,11 +75,13 @@ export function tuiCreateNumberMask({
             ? integer.slice(0, integerLimit + thousandSeparators.length)
             : integer;
         const integerCappedClean = integerCapped.replace(TUI_NON_DIGITS_REGEXP, '');
-        const leadingZerosMarchArray =
-            integerCappedClean.match(TUI_LEADING_ZEROES_REGEXP) || [];
-        const leadingZerosAmount =
-            leadingZerosMarchArray.length !== 0 ? leadingZerosMarchArray[0].length : 0;
-        const integerCappedZerosClean = integerCappedClean.replace(/^0+(?!\.|$)/, '');
+        const [leadingZerosMarch] = integerCappedClean.match(
```
Comment (splincode) [inline projects/core/utils/mask/create-number-mask.ts:78]: why `*March`? instead `*Match`

Comment (Mahakon) [reply projects/core/utils/mask/create-number-mask.ts]: fixed

**Review by shoom3301 — APPROVED** [review]

**Review by splincode — APPROVED** [review]

**Review by vladimirpotekhin — APPROVED** [review]

**Comment by waterplea** [comment]
Thank you!
